### PR TITLE
Refactor mapflags code

### PR DESCRIPTION
**Pull request by Emistry**
### Pull Request Prelude

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

- refactor mapflag related codes
- rearranged the codes based on mapflag index (easier for future to find and refer?)
- added missing mapflags.  
```
	mf_nomapchannelautojoin:    61
	mf_noknockback:             62
	mf_cvc:                     63
	mf_src4instance             64
```

**Issues addressed:** 
none

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

**Comment by 4144** [comment]
While you changed map flags, probably better remove map flags duplication.
Now at same time flags added into constants.conf and in C code.
Better remove them from constants.conf and add C enum values into script engine by ``script->set_constant``, this can reduce future errors.

**Comment by Emistry** [comment]
hmm should I keep the constants `MF_***` lowercase like the old time? or uppercase is more prefer?

**Comment by Kenpachi2k13** [comment]
I'd prefer upper case for the sake of consistency.

File: src/map/npc.c [hunk 1]
```diff
@@ -4777,32 +4777,38 @@ static const char *npc_parse_mapflag(const char *w1, const char *w2, const char
...
+		return strchr(start, '\n');// skip and continue
 	}
 
-	if (w4 && !strcmpi(w4, "off"))
+	if (w4 && strcmpi(w4, "off") == 0)
 		state = 0; //Disable mapflag rather than enable it. [Skotlex]
 
-	if (!strcmpi(w3, "nosave")) {
+	if (strcmpi(w3, "nomemo") == 0)
+		map->list[m].flag.nomemo = (state) ? 1 : 0;
```
Comment (Kenpachi2k13) [inline src/map/npc.c:4804]:
```suggestion
		map->list[m].flag.nomemo = (state != 0) ? 1 : 0;
```
Same for the other changes in file.

File: src/map/script.c [hunk 2]
```diff
@@ -13516,72 +13516,76 @@ static BUILDIN(getmapinfo)
 
 static BUILDIN(getmapflag)
 {
-	int16 m,i;
+	int16 m, i;
```
Comment (Kenpachi2k13) [inline src/map/script.c:13519]:
```suggestion
	int m;
	int i;
```

File: src/map/script.c [hunk 3]
```diff
@@ -13516,72 +13516,76 @@ static BUILDIN(getmapinfo)
...
 	const char *str;
 
-	str=script_getstr(st,2);
-	i=script_getnum(st,3);
+	str = script_getstr(st, 2);
+	i = script_getnum(st, 3);
```
Comment (Kenpachi2k13) [inline src/map/script.c:13523]: Join with declarations.

File: src/map/script.c [hunk 4]
```diff
@@ -13516,72 +13516,76 @@ static BUILDIN(getmapinfo)
...
-			case MF_NOCASHSHOP:         script_pushint(st,map->list[m].flag.nocashshop); break;
-			case MF_NOAUTOLOOT:         script_pushint(st, map->list[m].flag.noautoloot); break;
-			case MF_NOVIEWID:           script_pushint(st, map->list[m].flag.noviewid); break;
-			case MF_PAIRSHIP_STARTABLE: script_pushint(st, map->list[m].flag.pairship_startable); break;
-			case MF_PAIRSHIP_ENDABLE:   script_pushint(st, map->list[m].flag.pairship_endable); break;
-			case MF_NOSTORAGE:          script_pushint(st, map->list[m].flag.nostorage); break;
-			case MF_NOGSTORAGE:         script_pushint(st, map->list[m].flag.nogstorage); break;
+	if (m >= 0) {
+		switch (i) {
+			case MF_NOMEMO:                  script_pushint(st, map->list[m].flag.nomemo); break;
```
Comment (Kenpachi2k13) [inline src/map/script.c:13528]: Fix indentation. Cases should have same indentation level as `switch`. (See [style guide](https://github.com/HerculesWS/Hercules/wiki/Coding-Style#chapter-1-indentation).)

File: src/map/script.c (hunk 3 above)
Comment (Emistry) [reply src/map/script.c]: updated

File: src/map/script.c (hunk 2 above)
Comment (Emistry) [reply src/map/script.c]: updated

File: src/map/script.c (hunk 4 above)
Comment (Emistry) [reply src/map/script.c]: updated

File: src/map/npc.c (hunk 1 above)
Comment (Emistry) [reply src/map/npc.c]: updated

File: src/map/npc.c [hunk 5]
```diff
@@ -4777,32 +4777,38 @@ static const char *npc_parse_mapflag(const char *w1, const char *w2, const char
...
-		if (retval) *retval = EXIT_FAILURE;
-		return strchr(start,'\n');// skip and continue
+			filepath, strline(buffer, start - buffer), mapname, w1, w2, w3, w4);
+		if (retval) 
+			*retval = EXIT_FAILURE;
+		return strchr(start, '\n');// skip and continue
 	}
 
-	if (w4 && !strcmpi(w4, "off"))
+	if (w4 && strcmpi(w4, "off") == 0)
```
[the remark this answers is not in the document]
Comment (Emistry) [reply src/map/npc.c]: updated

File: db/re/item_db.conf [hunk 6]
```diff
@@ -47389,7 +47389,7 @@ item_db: (
 		else bonus2 bHPLossRate,888,5000;
 	">
 	OnUnequipScript: <"
-		if((Hp<=999) && !getmapflag(strcharinfo(PC_MAP),mf_pvp) && !getmapflag(strcharinfo(PC_MAP),mf_pvp_noparty) && !getmapflag(strcharinfo(PC_MAP),mf_pvp_noguild)) {
+		if((Hp<=999) && !getmapflag(strcharinfo(PC_MAP), MF_PVP) && !getmapflag(strcharinfo(PC_MAP), MF_PVP_NOPARTY) && !getmapflag(strcharinfo(PC_MAP), MF_PVP_NOGUILD)) {
```
[the remark this answers is not in the document]
Comment (Emistry) [reply db/re/item_db.conf:47392]: updated

File: db/re/item_db.conf [hunk 7]
```diff
@@ -44413,7 +44413,7 @@ item_db: (
 		bonus2 bHPLossRate,666,10000;
 	">
 	OnUnequipScript: <"
-		if((Hp<=999) && !getmapflag(strcharinfo(PC_MAP),mf_pvp) && !getmapflag(strcharinfo(PC_MAP),mf_pvp_noparty) && !getmapflag(strcharinfo(PC_MAP),mf_pvp_noguild)) {
+		if((Hp<=999) && !getmapflag(strcharinfo(PC_MAP), MF_PVP) && !getmapflag(strcharinfo(PC_MAP), MF_PVP_NOPARTY) && !getmapflag(strcharinfo(PC_MAP), MF_PVP_NOGUILD)) {
```
[the remark this answers is not in the document]
Comment (Emistry) [reply db/re/item_db.conf:44416]: updated

File: db/pre-re/item_db.conf [hunk 8]
```diff
@@ -33885,7 +33885,7 @@ item_db: (
 		bonus2 bHPLossRate,666,10000;
 	">
 	OnUnequipScript: <"
-		if((Hp <= 999) && !getmapflag(strcharinfo(PC_MAP),mf_pvp) && !getmapflag(strcharinfo(PC_MAP),mf_pvp_noparty) && !getmapflag(strcharinfo(PC_MAP),mf_pvp_noguild)) {
+		if((Hp <= 999) && !getmapflag(strcharinfo(PC_MAP), MF_PVP) && !getmapflag(strcharinfo(PC_MAP), MF_PVP_NOPARTY) && !getmapflag(strcharinfo(PC_MAP), MF_PVP_NOGUILD)) {
```
[the remark this answers is not in the document]
Comment (Emistry) [reply db/pre-re/item_db.conf:33888]: updated

File: npc/other/poring_war.txt [hunk 9]
```diff
@@ -356,10 +356,10 @@ OnPCKillEvent:
...
 
 OnInit:
-	removemapflag "poring_w02",mf_partylock;
-	removemapflag "poring_w02",mf_pvp;
-	removemapflag "poring_w02",mf_pvp_noguild;
-	removemapflag "poring_w02",mf_pvp_nocalcrank;
+	removemapflag "poring_w02", MF_PARTYLOCK;
+	removemapflag "poring_w02", MF_PVP;
+	removemapflag "poring_w02", MF_PVP_NOGUILD;
+	removemapflag "poring_w02", MF_PVP_NOCALCRANK;
```
[the remark this answers is not in the document]
Comment (Emistry) [reply npc/other/poring_war.txt]: updated

File: npc/custom/woe_controller.txt [hunk 10]
```diff
@@ -92,7 +92,7 @@ OnInit:
 	set .Size, getarraysize($WOE_CONTROL);
 	if (.AutoKick || .NoOwner)
 		for(set .@i,0; .@i<30; set .@i,.@i+1) {
-			setmapflag .Castles$[.@i], mf_loadevent;
+			setmapflag .Castles$[.@i], MF_LOADEVENT;
```
[the remark this answers is not in the document]
Comment (Emistry) [reply npc/custom/woe_controller.txt]: updated

File: src/map/atcommand.c [hunk 11]
```diff
@@ -4075,6 +4075,8 @@ ACMD(mapinfo)
 		strcat(atcmd_output, msg_fd(fd,1088)); // AllowKS |
 	if (map->list[m_id].flag.reset)
 		strcat(atcmd_output, msg_fd(fd,1089)); // Reset |
+	if (map->list[m_id].flag.src4instance)
```
[the remark this answers is not in the document]
Comment (Emistry) [reply src/map/atcommand.c]: updated

File: src/map/atcommand.c [hunk 12]
```diff
@@ -4100,6 +4102,11 @@ ACMD(mapinfo)
 		strcat(atcmd_output, msg_fd(fd, 1292)); // PrivateAirshipStartable |
 	if (map->list[m_id].flag.pairship_endable)
 		strcat(atcmd_output, msg_fd(fd, 1293)); // PrivateAirshipEndable |
+	if (map->list[m_id].flag.chsysnolocalaj)
+		strcat(atcmd_output, msg_fd(fd, 853)); // No Map Channel Auto Join |
+	if (map->list[m_id].flag.src4instance)
```
[the remark this answers is not in the document]
Comment (Emistry) [reply src/map/atcommand.c]: updated

File: src/map/atcommand.c [hunk 13]
```diff
@@ -4100,6 +4102,11 @@ ACMD(mapinfo)
 		strcat(atcmd_output, msg_fd(fd, 1292)); // PrivateAirshipStartable |
 	if (map->list[m_id].flag.pairship_endable)
 		strcat(atcmd_output, msg_fd(fd, 1293)); // PrivateAirshipEndable |
+	if (map->list[m_id].flag.chsysnolocalaj)
```
[the remark this answers is not in the document]
Comment (Emistry) [reply src/map/atcommand.c]: updated

File: src/map/atcommand.c [hunk 14]
```diff
@@ -7918,33 +7925,80 @@ ACMD(mapflag)
 } while(0)
 
 	char flag_name[100];
-	short flag=0,i;
+	short flag = 0, i;
```
[the remark this answers is not in the document]
Comment (Emistry) [reply src/map/atcommand.c]: updated

File: src/map/map.c [hunk 15]
```diff
@@ -4721,187 +4722,330 @@ static bool map_zone_mf_cache(int m, char *flag, char *params)
 	nullpo_retr(false, params);
 	Assert_retr(false, m >= 0 && m < map->count);
 
-	if (params[0] != '\0' && !strcmpi(params, "off"))
+	if (params[0] != '\0' && strcmpi(params, "off") == 0)
 		state = 0;
 
-	if (!strcmpi(flag, "nosave")) {
+	if (strcmpi(flag, "nomemo") == 0) {
+		if (state && map->list[m].flag.nomemo)
```
[the remark this answers is not in the document]
Comment (Emistry) [reply src/map/map.c]: updated

File: src/map/atcommand.c [hunk 16]
```diff
@@ -7918,33 +7925,80 @@ ACMD(mapflag)
...
+		CHECKFLAG(noviewid);
+
+		clif->message(sd->fd, " ");
+		clif->message(sd->fd, msg_fd(fd, 1312)); // Usage: "@mapflag monster_noteleport 1" (0=Off | 1=On)
+		clif->message(sd->fd, msg_fd(fd, 1313)); // Type "@mapflag available" to list the available mapflags.
 		return true;
 	}
-	for (i = 0; flag_name[i]; i++) flag_name[i] = TOLOWER(flag_name[i]); //lowercase
+
+	for (i = 0; flag_name[i]; i++)
```
[the remark this answers is not in the document]
Comment (Emistry) [reply src/map/atcommand.c]: updated

File: src/map/map.c [hunk 17]
```diff
@@ -5291,164 +5412,115 @@ static bool map_zone_mf_cache(int m, char *flag, char *params)
...
+		for (k = 0; k < len; k++) {
+			if (skill_name[k] == '\t') {
+				memcpy(modifier, &skill_name[k + 1], len - k);
 				skill_name[k] = '\0';
 				break;
 			}
 		}
 
-		if( modifier[0] == '\0' || !( skill_id = skill->name2id(skill_name) ) || atoi(modifier) < 1 || atoi(modifier) > USHRT_MAX ) {
+		if (modifier[0] == '\0' || !(skill_id = skill->name2id(skill_name)) || atoi(modifier) < 1 || atoi(modifier) > USHRT_MAX) {
```
[the remark this answers is not in the document]
Comment (Emistry) [reply src/map/map.c]: updated

File: src/map/npc.c [hunk 18]
```diff
@@ -4777,32 +4777,38 @@ static const char *npc_parse_mapflag(const char *w1, const char *w2, const char
...
 
 	// w1=<mapname>
-	if( sscanf(w1, "%31[^,]", mapname) != 1 )
+	if (sscanf(w1, "%31[^,]", mapname) != 1)
 	{
-		ShowError("npc_parse_mapflag: Invalid mapflag definition in file '%s', line '%d'.\n * w1=%s\n * w2=%s\n * w3=%s\n * w4=%s\n", filepath, strline(buffer,start-buffer), w1, w2, w3, w4);
-		if (retval) *retval = EXIT_FAILURE;
-		return strchr(start,'\n');// skip and continue
+		ShowError("npc_parse_mapflag: Invalid mapflag definition in file '%s', line '%d'.\n * w1=%s\n * w2=%s\n * w3=%s\n * w4=%s\n", filepath, strline(buffer, start - buffer), w1, w2, w3, w4);
+		if (retval)
```
[the remark this answers is not in the document]
Comment (Emistry) [reply src/map/npc.c]: updated

File: src/map/map.c [hunk 19]
```diff
@@ -5257,32 +5378,32 @@ static bool map_zone_mf_cache(int m, char *flag, char *params)
...
+		for (k = 0; k < len; k++) {
+			if (skill_name[k] == '\t') {
+				memcpy(modifier, &skill_name[k + 1], len - k);
 				skill_name[k] = '\0';
 				break;
 			}
 		}
 
-		if( modifier[0] == '\0' || !( skill_id = skill->name2id(skill_name) ) || !skill->get_unit_id( skill->name2id(skill_name), 0) || atoi(modifier) < 1 || atoi(modifier) > USHRT_MAX ) {
+		if (modifier[0] == '\0' || !(skill_id = skill->name2id(skill_name)) || !skill->get_unit_id(skill->name2id(skill_name), 0) || atoi(modifier) < 1 || atoi(modifier) > USHRT_MAX) {
```
[the remark this answers is not in the document]
Comment (Emistry) [reply src/map/map.c]: updated

File: src/map/npc.c [hunk 20]
```diff
@@ -4777,32 +4777,38 @@ static const char *npc_parse_mapflag(const char *w1, const char *w2, const char
...
 
-	nullpo_retr(strchr(start,'\n'), w1);
-	nullpo_retr(strchr(start,'\n'), w3);
+	nullpo_retr(strchr(start, '\n'), w1);
+	nullpo_retr(strchr(start, '\n'), w3);
 
 	// w1=<mapname>
-	if( sscanf(w1, "%31[^,]", mapname) != 1 )
+	if (sscanf(w1, "%31[^,]", mapname) != 1)
 	{
```
[the remark this answers is not in the document]
Comment (Emistry) [reply src/map/npc.c]: f (sscanf(w1, "%31[^,]", mapname) != 1) {

File: src/map/atcommand.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (Emistry) [reply src/map/atcommand.c:8033]: hmm seem like this changes will caused the build to fail.

File: src/map/atcommand.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (Emistry) [reply src/map/atcommand.c]: hmm seem like this changes will caused the build to fail.

File: db/re/item_db.conf (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kenpachi2k13) [reply db/re/item_db.conf:47392]: This one is still missing.

File: src/map/script.c [hunk 21]
```diff
@@ -14095,55 +14096,56 @@ static BUILDIN(setmapflag)
...
-		switch(i) {
-			case MF_NOMEMO:             map->list[m].flag.nomemo = 1; break;
-			case MF_NOTELEPORT:         map->list[m].flag.noteleport = 1; break;
-			case MF_NOSAVE:             map->list[m].flag.nosave = 1; break;
-			case MF_NOBRANCH:           map->list[m].flag.nobranch = 1; break;
-			case MF_NOPENALTY:          map->list[m].flag.noexppenalty = 1; map->list[m].flag.nozenypenalty = 1; break;
-			case MF_NOZENYPENALTY:      map->list[m].flag.nozenypenalty = 1; break;
+	if (m >= 0) {
+		switch (i) {
+			case MF_NOMEMO:                  map->list[m].flag.nomemo = 1; break;
```
Comment (Kenpachi2k13) [inline src/map/script.c:14101]: Wrong indentation. Cases should have the same indentation as the `switch()`.

File: src/map/script.c [hunk 22]
```diff
@@ -14075,15 +14077,14 @@ static int script_mapflag_pvp_sub(struct block_list *bl, va_list ap)
 
 static BUILDIN(setmapflag)
 {
-	int16 m,i;
+	int16 m, i;
```
Comment (Kenpachi2k13) [inline src/map/script.c:14080]: Please declare each variable in its own line.

File: src/map/script.c [hunk 23]
```diff
@@ -14095,55 +14096,56 @@ static BUILDIN(setmapflag)
...
+			case MF_NOTELEPORT:              map->list[m].flag.noteleport = 1; break;
+			case MF_NOSAVE:                  map->list[m].flag.nosave = 1; break;
+			case MF_NOBRANCH:                map->list[m].flag.nobranch = 1; break;
+			case MF_NOPENALTY:               map->list[m].flag.noexppenalty = 1; map->list[m].flag.nozenypenalty = 1; break;
+			case MF_NOZENYPENALTY:           map->list[m].flag.nozenypenalty = 1; break;
 			case MF_PVP:
 				map->list[m].flag.pvp = 1;
-				if( !battle_config.pk_mode ) {
-					map->foreachinmap(script->mapflag_pvp_sub,m,BL_PC);
+				if (!battle_config.pk_mode) {
```
Comment (Kenpachi2k13) [inline src/map/script.c:14109]:
```suggestion
				if (battle_config.pk_mode == 0) {
```

File: src/map/script.c [hunk 24]
```diff
@@ -14185,91 +14191,95 @@ static BUILDIN(setmapflag)
 
 static BUILDIN(removemapflag)
 {
-	int16 m,i;
+	int16 m, i;
```
Comment (Kenpachi2k13) [inline src/map/script.c:14194]: Please declare each variable in its own line.

File: src/map/script.c [hunk 25]
```diff
@@ -14185,91 +14191,95 @@ static BUILDIN(setmapflag)
...
-			case MF_NOMEMO:             map->list[m].flag.nomemo = 0; break;
-			case MF_NOTELEPORT:         map->list[m].flag.noteleport = 0; break;
-			case MF_NOSAVE:             map->list[m].flag.nosave = 0; break;
-			case MF_NOBRANCH:           map->list[m].flag.nobranch = 0; break;
-			case MF_NOPENALTY:          map->list[m].flag.noexppenalty = 0; map->list[m].flag.nozenypenalty = 0; break;
-			case MF_NOZENYPENALTY:      map->list[m].flag.nozenypenalty = 0; break;
-			case MF_PVP: {
+	if (m >= 0) {
+		switch (i) {
+			case MF_NOMEMO:                  map->list[m].flag.nomemo = 0; break;
```
Comment (Kenpachi2k13) [inline src/map/script.c:14203]: Wrong indentation. Cases should have the same indentation as the `switch()`.

File: src/map/atcommand.c [hunk 26]
```diff
@@ -3934,254 +3934,261 @@ ACMD(mapinfo)
...
+	clif->message(fd, msg_fd(fd, 1039)); // ------ Map Info ------
 
 	// count chats (for initial message)
 	chat_num = 0;
 	iter = mapit_getallusers();
 	for (pl_sd = BL_UCCAST(BL_PC, mapit->first(iter)); mapit->exists(iter); pl_sd = BL_UCCAST(BL_PC, mapit->next(iter))) {
-		if( pl_sd->mapindex == m_index ) {
-			if( pl_sd->state.vending )
+		if (pl_sd->mapindex == m_index) {
+			if (pl_sd->state.vending)
```
Comment (Kenpachi2k13) [inline src/map/atcommand.c:3961]:
```suggestion
			if (pl_sd->state.vending != 0)
```

File: src/map/atcommand.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (Kenpachi2k13) [reply src/map/atcommand.c:8033]: Yeah, my fault. It's `flag_name[i] != '\0'`.

File: src/map/script.c (hunk 21 above)
Comment (Nurkz) [reply src/map/script.c:14101]: But it's weird( mby change general typo rules to more c/c++ friendly? @Kenpachi2k13

**Comment by MishimaHaruna** [comment]
Resolved a few remaining loose ends and force pushed

**Review by MishimaHaruna — APPROVED** [review]

**Pull request merged**